### ISSUE TITLE
Authorization fails due to Authorization Header Lost on Redirect

### ISSUE THREAD

**Issue by Tasteful** [NuGet/Home#2521]
After the update to NuGet Version: 3.3.0.212 and Visual Studio VSIX Package Version: 3.4.2.830 password protected nuget feeds is not working.

When using command line I got the following error even that the credentials is stored in nuget.config file. And from inside VS the credential dialog is poping up all the time.

```
C:\>nuget list -source https://domain/nuget -verbosity detailed
Using credentials from config. UserName: myusername
Please provide credentials for: https://domain/nuget
UserName: System.InvalidOperationException: Cannot prompt for input in non-interactive mode.
   at NuGet.Common.Console.EnsureInteractive()
   at NuGet.Common.Console.ReadLine()
   at NuGet.ConsoleCredentialProvider.Get(Uri uri, IWebProxy proxy, Boolean isProxy, Boolean isRetry, Boolean nonInteractive, CancellationToken cancellationToken)
   at NuGet.Credentials.CredentialService.<GetCredentials>d__11.MoveNext()
```

When I use fiddler to check the request to server I can see that the authentication not is encrypted correct. When I decode the base64 decoded string I can see that the username is correct but the password in some encoded format (base64 also).

**Comment by Tasteful** [comment]
After resaved the feeds credential into nuget.config the command line utilitiy is working without asking for credentials. VS still asks for credentials.

**Comment by yishaigalatzer** [comment]
Can you share what server are you using, and how is authentication setup, so we can reproduce the issue?

**Comment by yishaigalatzer** [comment]
CC @pspill

**Comment by Tasteful** [comment]
I have setup a test-feed on http://nuget.product.workplace.nu/nuget that using the same basic authentication that my other feed (the only different is that this feed not using https).

User: nugetbug
Password: bugnuget

Add the feed into VS and try to browse for some packages.

**Comment by Tasteful** [comment]
the part that is used to parse the authentication header is the following.

```
        private string[] ParseAuthHeader(string authHeader)
        {
            // Check this is a Basic Auth header
            if (string.IsNullOrEmpty(authHeader) || !authHeader.StartsWith("Basic")) return null;

            // Pull out the Credentials with are seperated by ':' and Base64 encoded
            string base64Credentials = authHeader.Substring(6);
            string[] credentials = Encoding.ASCII.GetString(Convert.FromBase64String(base64Credentials)).Split(new[] { ':' });

            if (credentials.Length != 2 || string.IsNullOrEmpty(credentials[0]) || string.IsNullOrEmpty(credentials[0])) return null;

            // Okay this is the credentials
            return credentials;
        }
```

**Comment by yishaigalatzer** [comment]
Can you please also tell us what server is this?

**Comment by Tasteful** [comment]
It's an Windows 2012R2 server, nuget application server is from NuGet.Server.2.10.0-master-0002 with an http-module to validate password with local datastore. Everything was working untill the update that was installed on my computer 9th of April.

**Comment by yishaigalatzer** [comment]
Sounds good. We will debug in a few minutes.

As a side note I'd suggest updating to the latest nuget.server package. We have made significant performance improvements

**Comment by Tasteful** [comment]
Perfect with more performance improvements :)

**Comment by yishaigalatzer** [comment]
@alpaix discovered that if you add "/" to the end of the source url, things magically start working. We haven't figured out the exact reason this fails, but we see a 307 temporary redirect and after that comes in things fall apart. With adding the "/" things just work.

Could you confirm that this unblocks you at least, and we will continue investigating?

**Comment by Tasteful** [comment]
Good finding, I have updated my URLs to include the trailing / and now it works within VS also.

With nuget command line utility it works both with and without the trailing /.

**Comment by Nanako-Kbs** [comment]
Had same problem, adding trailing slash to the URL fixed it.

**Comment by scottwstone** [comment]
I have a problem with the NuGet Package Manager Extension for VS 2015 (v 3.4.2.830).  After updating, I get a prompt whenever trying to connect to a private repository hosted in IIS that requires Windows Authentication.  I can hit the URL fine from a browser, and connect from NuGet.exe, but both NuGet Package Manager and NuGet Package Explorer get 401s.  The Package Manager displays a login prompt, but continues to prompt even after valid credentials have been entered.  I have an older version of the extension installed in VS 2013 and it continues to work.  I have attempted to set up a new repository using the latest Nuget.Server assembly and it has the same issue.

**Comment by scottwstone** [comment]
Oh, and adding a trailing slash to the URL did not help.

**Comment by yishaigalatzer** [comment]
@scottwstone this issue is tracked here https://github.com/NuGet/Home/issues/2521

**Comment by yishaigalatzer** [comment]
@scottwstone can you please share a clear repro so we can validate this fix addresses your scenario as well?

**Comment by scottwstone** [comment]
@yishaigalatzer are you asking for a url to a repository or for details about the web server?

**Comment by yishaigalatzer** [comment]
Either will help

**Comment by scottwstone** [comment]
I am running a repository on Windows Server 2012 R2 server (IIS 8.5.9600.16384) using the NuGet Server package v 2.10.2 in an empty web project.  IIS is configured with Windows Authentication Enabled (response type = HTTP 401 Challenge) and all other forms of authentication disabled.  

My NuGet Package Manager -> Package Sources includes an entry named "Nuget-3.4.2" pointed at http://mydomain.com/NugetPrivate-3.4.2/nuget/.  When I:
1. open the Package Manager Console and select Nuget-3.4.2 from the Package Source dropdown, 
2. Type Get-Package -ListAvailable -Source Nuget-3.4.2

I get a login prompt.  I enter valid credentials, and get another login prompt.  Eventually, I hit cancel and Package Manager Console says "No packages found in the current package source."

If I open a command prompt and run: 

```
  nuget.exe -list -source Nuget-3.4.2 
```

I get a list of packages

**Comment by scottwstone** [comment]
I've done some more testing and it looks like my issue is related to something else on my network.  Please disregard.

**Comment by yishaigalatzer** [comment]
Thanks @scottwstone for saving us the trip down the rabbit hole :)

**Comment by scottwstone** [comment]
no problem.  I'll post more details about what was getting in the way if I can find them.  I worked around the issue by setting a hosts file redirect on my local machine to go straight to the ip of the server that hosts the nuget repository.

**Comment by joelverhagen** [comment]
Here is a new build with some improvements in this area. Note that this is pre-release build. Could you try this VSIX out in your own Visual Studio 2015 and see if it resolves the problem?
https://www.myget.org/F/nuget-vsix/vsix/NuGet.0d421874-a3b2-4f67-b53a-ecfce878063b-3.4.3.841.vsix

**Comment by Tasteful** [comment]
I have now tried the VSIX and for me it seams to work with and without the trailing slash.

Good work!

**Comment by vcaraulean** [comment]
I had same problem with authentication on local nuget repository. It was even worse, it prevented from installing any packages for any sources, including nuget.org.

I've installed pre-release build and it seems to be working now. Please, get it out of the door, we have few teams here that are using the local package source. I cannot tell all of them to install a pre-release package.

**Comment by vcaraulean** [comment]
@joelverhagen after installing this pre-release, will VS install the released version when it comes out? Or I'll have to uninstall pre-release and install new version manually?

**Comment by yishaigalatzer** [comment]
You will have to uninstall it. We are working on releasing it. It would be a few more days to finish other fixes and test

**Comment by scottwstone** [comment]
The 401 is resolved for me too.  Thanks!

**Comment by Tasteful** [comment]
@alpaix I think this issue not is fully solved. I can now list the packages when using the private authenticated feed, but can download them during package restore in the yellow bar in package manager console that saying that some packages is missing.

VSIX Version 3.4.3.855 (Also tested with version 3.5.0.1236 and the same issue)

```
GET https://nuget-release.test/nuget/ HTTP/1.1
user-agent: NuGet Client V3/3.4.3.0 (Microsoft Windows NT 10.0.10586.0, VS Enterprise/14.0)
Host: nuget-release.test
Accept-Encoding: gzip, deflate


HTTP/1.1 401 Unauthorized
Content-Length: 1293
Content-Type: text/html
WWW-Authenticate: Basic realm="Use your login for test"
Server: Microsoft-IIS/8.5 Microsoft-HTTPAPI/2.0
X-Powered-By: ASP.NET
X-Robots-Tag: noindex, nofollow
Date: Mon, 02 May 2016 06:58:22 GMT
------------------------------------------------------------------

GET https://nuget-release.test/nuget/ HTTP/1.1
user-agent: NuGet Client V3/3.4.3.0 (Microsoft Windows NT 10.0.10586.0, VS Enterprise/14.0)
Accept-Encoding: gzip, deflate
Authorization: Basic [correct authentication token]
Host: nuget-release.test


HTTP/1.1 200 OK
Cache-Control: no-cache
Content-Length: 406
Content-Type: application/xml;charset=utf-8
Server: Microsoft-IIS/8.5 Microsoft-HTTPAPI/2.0
DataServiceVersion: 1.0;
X-AspNet-Version: 4.0.30319
X-Powered-By: ASP.NET
X-Robots-Tag: noindex, nofollow
Date: Mon, 02 May 2016 06:58:22 GMT
------------------------------------------------------------------

GET https://nuget-release.test/nuget/Packages(Id='Test.Utilities',Version='5.0.0') HTTP/1.1
Accept: application/atom+xml, application/xml
user-agent: NuGet Client V3/3.4.3.0 (Microsoft Windows NT 10.0.10586.0, VS Enterprise/14.0)
Host: nuget-release.test
Accept-Encoding: gzip, deflate


HTTP/1.1 401 Unauthorized
Content-Length: 1293
Content-Type: text/html
WWW-Authenticate: Basic realm="Use your login for test"
Server: Microsoft-IIS/8.5 Microsoft-HTTPAPI/2.0
X-Powered-By: ASP.NET
X-Robots-Tag: noindex, nofollow
Date: Mon, 02 May 2016 06:58:22 GMT
------------------------------------------------------------------

GET https://nuget-release.test/nuget/Packages(Id='test.Utilities',Version='5.0.0') HTTP/1.1
Accept: application/atom+xml, application/xml
user-agent: NuGet Client V3/3.4.3.0 (Microsoft Windows NT 10.0.10586.0, VS Enterprise/14.0)
Accept-Encoding: gzip, deflate
Authorization: Basic [bad authentication token ([username]:[base64 string that not is the password])]
Host: nuget-release.test


HTTP/1.1 401 Unauthorized
Content-Length: 1293
Content-Type: text/html
WWW-Authenticate: Basic realm="Use your login for test"
Server: Microsoft-IIS/8.5 Microsoft-HTTPAPI/2.0
X-Powered-By: ASP.NET
X-Robots-Tag: noindex, nofollow
Date: Mon, 02 May 2016 06:58:22 GMT
------------------------------------------------------------------
```

**Comment by Tasteful** [comment]
@alpaix it's not the same issue. I will create another issue for that one.

**Comment by yishaigalatzer** [comment]
@Tasteful before proceeding with a new bug please pick up 3.4.4 and let us know.

**Comment by SarathVakacharla** [comment]
I have this problem with 2.9 Version of Nuget with VS 2013

---

Microsoft Visual Studio Ultimate 2013
Version 12.0.31101.00 Update 4
Microsoft .NET Framework
Version 4.6.01038
Installed Version: Ultimate
NuGet Package Manager   2.9.0.780

---

Will there be a fix made for VS 2013?

**Comment by MartinJohns** [comment]
@alpaix Will there be a fix made for VS2013? I'm having the same issue with the latest NuGet tools.

**Comment by yishaigalatzer** [comment]
There is no plan to update 2013 with authentication pieces.
